cmd/xorgen/main.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"github.com/saylorsolutions/gocryptx/cmd/xorgen/internal/tmpl"
+	flag "github.com/spf13/pflag"
+	"io"
+	"os"
+	"strings"
+)
+
+func main() {
+	var (
+		helpFlag     bool
+		exposedFlag  bool
+		compressFlag bool
+	)
+	flags := flag.NewFlagSet("xorgen", flag.ContinueOnError)
+	flags.BoolVarP(&helpFlag, "help", "h", false, "Prints this usage information.")
+	flags.BoolVarP(&exposedFlag, "exposed", "E", false, "Make the unscreen method exposed from the file. It's recommended to only expose from within an internal package.")
+	flags.BoolVarP(&compressFlag, "compressed", "c", false, "Payload should be gzip compressed when embedded, which includes a checksum to help prevent tampering.")
+	flags.Usage = func() {
+		fmt.Printf(`
+xorgen generates code to embed XOR obfuscated (and optionally compressed) data by generating a *.go file based on the input file. This pairs well with go:generate comments.
+The name of the generated Go file will be based on the name of the input file, replacing characters that match the regex pattern [^a-zA-Z0-9_] with "_".
+For example, given a file called super-secret.txt, a Go file will be created in the current directory called super_secret_txt.go, containing a method called unscreenSuper_secret_txt.
+See the -E flag below to make it an exposed method, and make sure you review the SECURITY notes below if you're unfamiliar with XOR screening.
+
+USAGE:  xorgen FILE [KEY]
+
+Note: If a key argument is given, it will be used with offset 0.
+
+ARGS:
+    FILE is the input file to be embedded.
+    KEY is optional and may be specified to override secure random generation behavior.
+
+FLAGS:
+%s
+SECURITY:
+    This is not encryption, this is obfuscation, and they are very different things!
+XOR screening is intended to hide embedded data from passive binary analysis only, since XOR screening is easily reversible.
```

Comment:
*relatively* easily reversible.
Sizeable keys make it more difficult. Should add a note that compression makes it a bit easier to recover at least part of the XOR key once the data is found in the binary, since the gzip header is rather predictable. However, this is mostly irrelevant since the key is stored right next to it. So it's worth noting, but not really germane to the real threat model.